### PR TITLE
[MM-14381] Fix brittle MFA Login Successful Test

### PR DESCRIPTION
**Pull request by DSchalla**
#### Summary
The PR fixes the brittle `TestUserLoginMFAFlow/WithCorrectMFA` test. It was validated using:
`SECONDS=0; while go test ./... -run 'TestUserLoginMFAFlow/WithCorrectMFA'; do echo "PASSED - $SECONDS seconds elapsed"; done` with a runtime 1000s+ without any failures.

Since this is no guarantee, @lieut-data since you were able to replicate the races locally I'd appreciate a second test on your machine eventually to make sure that we don't introduce an unstable test once again.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-14381

### CODE REVIEW

File: api4/user_test.go [hunk 1]
```diff
@@ -2068,8 +2069,19 @@ func TestUserLoginMFAFlow(t *testing.T) {
 	})
 
 	t.Run("WithCorrectMFA", func(t *testing.T) {
-		t.Skip("Skipping test that fails randomly.")
-		code := dgoogauth.ComputeCode(secret.Secret, time.Now().UTC().Unix()/30)
+		secret3, err := th.App.GenerateMfaSecret(th.BasicUser.Id)
```
Comment (lieut-data) [inline api4/user_test.go]: Nit, can we just call this `secret` since everything is scoped to this function?

File: api4/user_test.go [hunk 2]
```diff
@@ -2068,8 +2069,19 @@ func TestUserLoginMFAFlow(t *testing.T) {
...
+		// Fake user has MFA enabled
+		if result := <-th.Server.Store.User().UpdateMfaActive(th.BasicUser.Id, true); result.Err != nil {
+			t.Fatal(result.Err)
+		}
+
+		if result := <-th.Server.Store.User().UpdateMfaSecret(th.BasicUser.Id, secret3.Secret); result.Err != nil {
+			t.Fatal(result.Err)
+		}
+
+		code := dgoogauth.ComputeCode(secret3.Secret, time.Now().UTC().Unix()/30)
```
Comment (lieut-data) [inline api4/user_test.go]: Fundamentally, I'm guessing that the race seems to arise from the fact that `time.Now()` changes between here and the code downstream from `th.Client.LoginWithMFA` below. Is our implementation such that we don't allow for overlap of codes? If so, I imagine this impacts real users from time to time as well. If we accepted prev, current, or next codes this problem would go away entirely, but I don't know if that's best practice or not.

Comment (DSchalla) [reply api4/user_test.go]: We currently support a window size of 3 (https://github.com/mattermost/mattermost-server/blob/master/services/mfa/mfa.go#L87) which should cover that, I am really unsure where the race is coming from currently. I suppose having multiple MFA providers where we could mock the implementation would be beneficial (similar to supporting Yubikey etc), but that would be a bigger refactor. I'll see when I can find for this again to retest, thanks for taking the time to test it Jesse!

File: api4/user_test.go [hunk 3]
```diff
@@ -2068,8 +2069,19 @@ func TestUserLoginMFAFlow(t *testing.T) {
...
 
 		user, resp := th.Client.LoginWithMFA(th.BasicUser.Email, th.BasicUser.Password, strconv.Itoa(code))
```
Comment (lieut-data) [inline api4/user_test.go]: @DSchalla, I think I've uncovered the core issue. dgoogauth's `ComputeCode` returns `int`, but it's Authenticate expects a `string`, and then validates the length as 6 (or 8 characters). After tracing the code to verify, I observed that this test fails whenever the string representation of the password is supposed to have a leading `zero`, but our `strconv.Itoa(code)` in the test discards same.

The library's own test code had exactly the same issue: https://github.com/dgryski/dgoogauth/commit/f176097db5de17a6c0f2f0d74abf9d38340fdd60.

As such we, can fix this test as follows:
```suggestion
		user, resp := th.Client.LoginWithMFA(th.BasicUser.Email, th.BasicUser.Password, fmt.Sprintf("%06d", code))
```

I'm unable to reproduce the issue with the changes above.

Comment (DSchalla) [reply api4/user_test.go]: :tada: Amazing find! This explains why I was never able to trigger it, considering the chance of a leading zero is not *that* high :-) Thanks a lot for the proposed fix and digging that deep, happy we got that fixed and can test the 2FA flow now properly.

**Pull request merged**